Bottom: 34c82ea48d9c77b760f36f379648c0a3a627ff83
Top:    2c7b12df4164653463aa74aaed6cd68b9a690b88
Author: Ian Kent <ikent@redhat.com>
Date:   2012-05-15 11:24:25 +0800

autofs-5.0.5 - dont start readmap unless ready

State transitions should only be made when an autofs point goes
to state ST_READY or when a new task is added to an empty task
queue.


---

diff --git a/daemon/state.c b/daemon/state.c
index 3c88335..f9bc240 100644
--- a/daemon/state.c
+++ b/daemon/state.c
@@ -816,14 +816,13 @@ done:
 		new = st_alloc_task(ap, state);
 		if (new)
 			list_add(&new->list, head);
+		/* Added to empty state queue, kick state machine */
+		signaled = 1;
+		status = pthread_cond_signal(&cond);
+		if (status)
+			fatal(status);
 	}
 
-	/* Added task, encourage state machine */
-	signaled = 1;
-	status = pthread_cond_signal(&cond);
-	if (status)
-		fatal(status);
-
 	return 1;
 }
 
@@ -1129,6 +1128,10 @@ static void *st_queue_handler(void *arg)
 				task = list_entry(p, struct state_queue, list);
 				p = p->next;
 
+				/* Task may have been canceled before it started */
+				if (!task->thid && task->cancel)
+					goto remove;
+
 				if (!task->busy) {
 					/* Start a new task */
 					task->busy = 1;
